Mathlib/Data/Matrix/Rank.lean

```diff
@@ -18,17 +18,105 @@ This definition does not depend on the choice of basis, see `Matrix.rank_eq_finr
+
+universe ul um un uo uR
+variable {l : Type ul} {m : Type um} {n : Type un} {o : Type uo} {R : Type uR}
+
+section Infinite
+
+variable [Semiring R] {A : Matrix m n R}
+
+/-- The rank of a matrix, defined as the dimension of its column space, as a cardinal. -/
+noncomputable def cRank (A : Matrix m n R) : Cardinal := Module.rank R <| span R <| range Aᵀ
+
+lemma cRank_toNat_eq_finrank (A : Matrix m n R) :
+    A.cRank.toNat = Module.finrank R (span R (range Aᵀ)) := rfl
+
+lemma cRank_mono_col (A : Matrix n o R) (c : m → o) :
+    (A.submatrix id c).cRank ≤ A.cRank := by
+  apply Submodule.rank_mono <| span_mono ?_
+  rintro _ ⟨x, rfl⟩
+  exact ⟨c x, rfl⟩
+
+lemma lift_cRank_submatrix_le
+    (A : Matrix n o R) (r : l → n) (c : m → o) :
+    lift.{un} (A.submatrix r c).cRank ≤ lift.{ul} A.cRank := by
+  trans lift.{un} <| lift.{ul} (A.submatrix r id).cRank
+  · refine Cardinal.lift_monotone ?_
+    trans lift.{ul} ((A.submatrix r id).submatrix id c).cRank
+    · rw [@lift_id'.{ul, uR}, submatrix_submatrix]
+      simp
+    · refine Cardinal.lift_monotone <| Submodule.rank_mono <| span_mono ?_
+      rintro _ ⟨x, rfl⟩
+      exact ⟨c x, rfl⟩
+  · rw [Cardinal.lift_lift]
+    let f : (_ → R) →ₗ[R] (_ → R) := LinearMap.funLeft R R r
+    have h_eq : Submodule.map f (span R (range Aᵀ)) = span R (range (A.submatrix r id)ᵀ) := by
+      rw [LinearMap.map_span, ← image_univ, image_image, transpose_submatrix]
+      aesop
+    rw [cRank, ← h_eq]
+    have hwin := lift_rank_map_le f (span R (range Aᵀ))
+    simp_rw [← lift_umax] at hwin ⊢
+    exact hwin
+
+/-- A copy of `lift_cRank_submatrix_le` for when `l` and `n` are in the same universe. -/
+lemma cRank_submatrix_le.{uln} {l n : Type uln}
+    (A : Matrix n o R) (r : l → n) (c : m → o) :
+    (A.submatrix r c).cRank ≤ A.cRank := by
+  simpa using lift_cRank_submatrix_le A r c
+
+lemma cRank_le_card_row [StrongRankCondition R] [Fintype m] (A : Matrix m n R) :
+    A.cRank ≤ Fintype.card m :=
+  (Submodule.rank_le (span R (range Aᵀ))).trans <| by rw [rank_fun']
+
+lemma cRank_le_card_col [StrongRankCondition R] [Fintype n] (A : Matrix m n R) :
+    A.cRank ≤ Fintype.card n :=
+  (rank_span_le ..).trans <| by simpa using Cardinal.mk_range_le_lift (f := Aᵀ)
+
+/-- The rank of a matrix, defined as the dimension of its column space, as a term in `ℕ∞`. -/
+noncomputable def eRank (A : Matrix m n R) : ℕ∞ := A.cRank.toENat
+
+lemma eRank_toNat_eq_finrank (A : Matrix m n R) :
+    A.eRank.toNat = Module.finrank R (span R (range Aᵀ)) :=
+  toNat_toENat ..
+
+lemma eRank_submatrix_le (A : Matrix n o R) (r : l → n) (c : m → o) :
+    (A.submatrix r c).eRank ≤ A.eRank := by
+  obtain hle | hlt := le_or_lt aleph0 (A.submatrix id c).cRank
+  · simp [eRank, toENat_eq_top.2 <| hle.trans <| A.cRank_mono_col c]
+  refine le_trans ?_ <| OrderHomClass.mono _ <| A.cRank_mono_col c
+  simpa using (toENat_le_iff_of_lt_aleph0 (by simpa)).2 <|
+    (A.submatrix id c).lift_cRank_submatrix_le r id
```

Comment:
Do we still need to go via the `id` versions here now that the cRank version is generalized?